test/python/passmanager/__init__.py

```diff
@@ -0,0 +1,77 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Pass manager test cases."""
+
+import io
+import re
+from itertools import zip_longest
+from logging import StreamHandler, getLogger
+
+from qiskit.passmanager import PassManagerState, WorkflowStatus, PropertySet
+from qiskit.test import QiskitTestCase
+
+
+class PassManagerTestCase(QiskitTestCase):
+    """Test case for the pass manager module."""
+
+    def setUp(self):
+        super().setUp()
+
+        self.output = io.StringIO()
+        self.state = PassManagerState(
+            workflow_status=WorkflowStatus(),
+            property_set=PropertySet(),
+        )
+
+        logger = getLogger()
+        self.addCleanup(logger.setLevel, logger.level)
+
+        logger.setLevel("DEBUG")
+        logger.addHandler(StreamHandler(self.output))
+
+    def assertLogEqual(self, func, expected_lines, *args, exception_type=None):
+        """Execute provided function and verify logger.
+
+        Args:
+            func (Callable): Function to test.
+            expected_lines (List[str]): Expected log output.
+            args (Any): Arguments to the function.
+            exception_type (Type): Optional. Expected exception for error handling.
+
+        Returns:
+            Any: Output values from the function.
+        """
+        if exception_type:
+            self.assertRaises(exception_type, func, *args)
+            out = None
+        else:
+            out = func(*args)
+
+        self.output.seek(0)
+        recorded_lines = [line.rstrip() for line in self.output.readlines()]
+        for i, (expected, recorded) in enumerate(zip_longest(expected_lines, recorded_lines)):
+            expected = expected or ""
+            recorded = recorded or ""
+            if not re.fullmatch(expected, recorded):
+                raise AssertionError(
+                    f"Log didn't match. Mismatch found at line #{i}.\n\n"
+                    f"Expected:\n{self._format_log(expected_lines)}\n"
+                    f"Recorded:\n{self._format_log(recorded_lines)}"
+                )
+        return out
+
```

Comment:
Is this code copied from somewhere?  The exception handling within this function shouldn't be necessary; wrapping a "standard" call to this function in the `self.assertRaises` context manager in the regular test should all work correctly.

It'd also likely be better if this function was a context-manager wrapper around `unittest.assertLogs` - it's much easier to read tests written in context-manager form than "exploded function call" (`assertRaises(f, arg0, arg1, ...)`).

Reply:
Right, but if they're both used as context managers, the clean-up code from both (the bit that contains the assertions) will still run in both correctly; we can use the context-manager stack to manage things arbitrarily, rather than needing hard-coded functions.

Reply:
In very very very rough form, what I'm meaning is something like:
```python
import contextlib

class TestCase:
    @contextlib.contextmanager
    def assertLogContains(self, expected_lines):        
        with self.assertLogs() as cm:
            yield cm
        recorded_lines = cm.output
        for i, (expected, recorded) in enumerate(zip(expected_lines, recorded_lines)):
            # ... your existing handling
```

then I think it should work in test cases as:
```python
def test_logs_while_raising(self):
    with self.assertLogContains(["blah blah"]):
        getLogger().log("blah blah")
        with self.assertRaises(Exception):
            raise Exception
```
and that test would pass.

Reply:
(possibly the log-checking code should be in a `finally` block, so there's no spurious passes if a test author accidentally puts the `assertLogContains` and `assertRaises` calls the wrong way round)

Reply:
Done in [846e416](https://github.com/Qiskit/qiskit/pull/11178/commits/846e41631a57a72216268f70dac43b394f10d4f0). I needed some ugly hack because of [this logic](https://github.com/python/cpython/blob/f4b5588bde656d8ad048b66a0be4cb5131f0d83f/Lib/unittest/_log.py#L31).

Reply:
Hmmm it didn't work. The default formatter is bit ugly but maybe okey.
```
Expected:
#00: Pass: TaskC - (\d*\.)?\d+ \(ms\)
#01: Pass: TaskB - (\d*\.)?\d+ \(ms\)

Recorded:
#00: INFO:qiskit.passmanager.base_tasks:Pass: TaskA - 0.00596 (ms)
#01: INFO:qiskit.passmanager.base_tasks:Pass: TaskB - 0.00691 (ms)
```

Reply:
Oh, that's unfortunately gross of the standard library.  Unfortunately, I think we're going to have problems with the private method: the `unittest/_log.py` was only split out from `unittest/case.py` in Python 3.9, so if we're going to go the way of accessing private internals, we'll need to have an import catch for Python 3.8 to pull it from `unittest.case` instead.   The actual private code didn't meaningfully change, though, so it'll just about hold up.

Is there any way we can write the test enforcing the default logging format, so we can avoid needing the private access?